Use the net/http library for HTTP requests

ironcladlou opened the pull request:
Replace curl usage with the net/http library. Support request retries
and rubust response handling.

Fix https://bugzilla.redhat.com/show_bug.cgi?id=1196617.

ironcladlou committed:
Use the net/http library for HTTP requests

Replace curl usage with the net/http library. Support request retries
and rubust response handling.

Fix https://bugzilla.redhat.com/show_bug.cgi?id=1196617.

ironcladlou commented:
This PR is still a WIP.

ironcladlou commented:
cc @detiber @a13m @jwhonce

ironcladlou commented:
Should each request to Jenkins be retried forever? If not, for how long? Are there any response codes we know for a fact at a point in time can be considered immediately fatal?

ironcladlou commented:
cc @sferich888 @brenton @sosiouxme

detiber commented:
@ironcladlou I would think we could consider the 4xx errors as fatal. Probably any 5xx errors other than 502, 503, and 504 as well.

ironcladlou commented:
Let's turn it around: what's the inclusive list of HTTP codes which should trigger a retry? 502/503/504?

detiber commented:
@ironcladlou I believe that to be the case, yes. That would cover Proxy errors (gear not available), Service Unavailable (gear overwhelmed) and Gateway timeout (node temporarily unavailable)

ironcladlou commented:
Refactored the code further and added comprehensive error handling and reporting. I think this is in much better shape than before. I need some code review before merging.

jwhonce commented on cartridges/openshift-origin-cartridge-jenkins-client/bin/jenkins_build:
```diff
@@ -2,37 +2,104 @@
+def try_request(type, uri_string, label)
+  response = nil
+  try_forever(label) do
+    http, uri = make_http(uri_string)
```
I think make_http() should be outside the try_forever() loop or you'll be stuck in the loop should the parse fail.

ironcladlou replied:
Good catch... I moved everything but the actual request out of the try loop.

ironcladlou commented:
[test]

ironcladlou commented:
@jwhonce give this one more review?

brenton commented on cartridges/openshift-origin-cartridge-jenkins-client/bin/jenkins_build:
```diff
@@ -2,37 +2,105 @@
+  try_forever(label) do    
+    response = http.request(request)
+    if RETRYABLE_ERRORS.include?(response.code)
+      raise "#{label}: HTTP #{response.code} #{response.message}"
```
I don't think we're printing this exception message anywhere.  It could be useful to do so.

ironcladlou replied:
In this context, I opted to eat it during retries and print periods to indicate progress since the errors are expected.... should I dump the message amongst the progress ellipses every second?

brenton replied:
We definitely wouldn't want this being returned to the client.  I thought maybe there was another log available.  It just looked odd to set a message that nothing ever read.

jwhonce commented on cartridges/openshift-origin-cartridge-jenkins-client/bin/jenkins_build:
```diff
@@ -45,62 +113,113 @@ def is_cancelled
-  status_code = `curl -s -w %{http_code} --output /dev/null -X POST --insecure https://${JENKINS_USERNAME}:${JENKINS_PASSWORD}@#{@hostname}#{@job_url}/build`
-  unless ['302', '200', '201'].include? status_code
-    puts "ERROR - Couldn't schedule job"
-    exit 292
```
I verified this exit code is _not_ on the broker's categorized list, so dropping it should not be an issue.

jwhonce commented:
LGTM :+1:

ironcladlou commented:
Thanks. Let's [merge] and let QE take it for a spin.
